### PR TITLE
mpi: split mpir implementation from mpi functions

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

The goal is to leave all top-level mpi functions bare boiler-plates, to facilitate script generation. Reference: #4846.
* [x] mpi/attr   - PR #4861 
* [x] mpi/coll
* [x] mpi/comm
* [x] mpi/datatype  
* [x] mpi/errhan  
* [x] mpi/group  
* [x] mpi/info  
* [x] mpi/init  
* [x] mpi/misc  - minor clean up
* [x] mpi/pt2pt - mostly direct `MPID` calls
* [x] mpi/request  
* [x] mpi/rma  - all direct `MPID` calls
* [x] mpi/spawn  
* [x] mpi/timer  - none needed
* [x] mpi/topo
TODO separately -
* [ ] mpi/misc/gpu.c - MPIX functions
* [ ] mpi_t
* [ ] mpi/romio  

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
Some of the MPIR functions are not inside the `#ifndef MPICH_MPI_FROM_PMPI` block, which should result in multiple definition of these functions. Why didn't it cause issue on mac?

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
This PR touches a lot of code so it is a daunting task to do a thorough review. I have added @shintaro-iwasaki to help. Each reviewer may choose to only review portion of the commits -- just make that clear in comments.

File: src/mpi/coll/barrier.c [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+/* -- End Profiling Symbol Block */
+
+/* Define MPICH_MPI_FROM_PMPI if weak symbols are not supported to build
+   the MPI routines */
+#ifndef MPICH_MPI_FROM_PMPI
+#undef MPI_Barrier
+#define MPI_Barrier PMPI_Barrier
+#endif
+
+
```
Comment (shintaro-iwasaki) [inline src/mpi/coll/barrier.c:28]: Is this series of newlines intentional? Even if the original code was like this, do we want to keep consecutive newlines even after reconstruction?

Comment (hzhou) [reply src/mpi/coll/barrier.c:28]: No, not intentional. It is a left-over from script executions. I'll make a note and update it later. On the hand, the plan next is remove these boiler-plates and generate them by script, so the whitespace issues may not be relevant soon.

Comment (shintaro-iwasaki) [reply src/mpi/coll/barrier.c:28]: Okay, then I will ignore.

**Comment by shintaro-iwasaki** [comment]
I quickly go through most of the external functions.  I could not check 16K lines, but basically it looks good to me. If it passes the tests, I believe this change does not break functionality.

A few questions:

1. Is it okay to keep `MPII_` functions in the same file for this #4846 purpose? (e.g., `MPII_Type_set_attr()` https://github.com/hzhou/mpich/blob/2010_mpi_coll/src/mpi/attr/type_set_attr.c)
Others are:
`MPII_Type_get_attr()`, `MPII_Type_set_attr()`, `MPII_Comm_get_attr()`, `MPII_Comm_set_attr()`, `MPII_Win_get_attr()`, `MPII_Win_set_attr()`

2. `misc/gpu.c` defines two external functions (although they are `MPIX_`)

3. If top-level functions become "bare boiler-plates" especially for scripts, do we want to leave files that define "public functions" and files that implement utility/MPIR etc in the same directory?

**Comment by hzhou** [comment]
> I quickly go through most of the external functions. I could not check 16K lines, but basically it looks good to me. If it passes the tests, I believe this change does not break functionality.
> 
> A few questions:
> 
>     1. Is it okay to keep `MPII_` functions in the same file for this #4846 purpose? (e.g., `MPII_Type_set_attr()` https://github.com/hzhou/mpich/blob/2010_mpi_coll/src/mpi/attr/type_set_attr.c)
>        Others are:
>        `MPII_Type_get_attr()`, `MPII_Type_set_attr()`, `MPII_Comm_get_attr()`, `MPII_Comm_set_attr()`, `MPII_Win_get_attr()`, `MPII_Win_set_attr()`

`MPII_` is not really the correct prefix since it is exported and being called externally (by Fortran bindings), but let's not to change it for now. 

These `MPII_` functions are really `MPI_` functions with one extra parameter that will be passed-through. They are all boiler-plates and will be generated by the script. That's why they are kept in the mpi-level files.

> 
>     2. `misc/gpu.c` defines two external functions (although they are `MPIX_`)

Thanks. I'll add a todo check box. I'll either update it or leave it to another PR (along with MPI_T and ROMIO).

> 
>     3. If top-level functions become "bare boiler-plates" especially for scripts, do we want to leave files that define "public functions" and files that implement utility/MPIR etc in the same directory?

We can easily move the generated files somewhere else rather than moving the current utility code out. Let's worry about that later.

**Review by shintaro-iwasaki — APPROVED** [review]
As far as I checked the definitions of `MPI_` functions, this PR looks good to me.

**Comment by hzhou** [comment]
> As far as I checked the definitions of `MPI_` functions, this PR looks good to me.

Thanks @shintaro-iwasaki

File: src/include/mpir_misc.h [hunk 2]
```diff
@@ -56,6 +56,11 @@ extern const char MPII_Version_F77[] MPICH_API_PUBLIC;
 extern const char MPII_Version_FC[] MPICH_API_PUBLIC;
 extern const char MPII_Version_custom[] MPICH_API_PUBLIC;
 
+int MPIR_Init(int *argc, char ***argv);
+int MPIR_Init_thread(int *argc, char ***argv, int user_required, int *provided);
+int MPIR_Finalize(void);
+int MPIR_Abort_impl(MPIR_Comm * comm_ptr, int errorcode);
```
Comment (raffenet) [inline src/include/mpir_misc.h:62]: just curious, why `_impl` for abort, but none of the other functions?

Comment (hzhou) [reply src/include/mpir_misc.h:62]: This is the current inconsistency. My plan is to consistently use `_impl`, will do it in a separate PR.

**Review by raffenet — APPROVED** [review]
This is moving us in the right direction. Obviously hard to review closely, but we can fix bugs and such as we find them.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Pull request merged**